plugins/camera/camera_impl.cpp

```diff
@@ -0,0 +1,739 @@
+
+    for (auto parameter : parameters) {
+        // LogDebug() << "parameter to request: " << parameter.first;
+        if (parameter.second->type == CameraDefinition::Type::UINT32) {
+            _parent.get_param_ext_int_async(parameter.first,
+                                            std::bind(&CameraImpl::receive_int_param, this, parameter.first, _1, _2));
+        } else if (parameter.second->type == CameraDefinition::Type::FLOAT) {
+            _parent.get_param_ext_float_async(parameter.first,
+                                              std::bind(&CameraImpl::receive_float_param, this, parameter.first, _1, _2));
+        }
+    }
+
+}
+
+bool CameraImpl::get_possible_settings(std::map<std::string, std::string> &settings)
+{
+    if (!_camera_definition) {
+        LogWarn() << "Error: no camera definition available yet";
+        return false;
+    }
+
+    CameraDefinition::parameter_map_t parameters;
+    _camera_definition->get_parameters(parameters, true);
+
+    for (auto &parameter : parameters) {
+        if (parameter.first.compare("CAM_MODE") == 0) {
+            // We ignore the mode param.
+            continue;
+        }
+
+        settings.insert(
+            std::pair<std::string, std::string>(
+                parameter.first, parameter.second->description));
+    }
+
+    return true;
+}
+
+bool CameraImpl::get_possible_options(const std::string &setting_name,
+                                      std::vector<std::string> &options)
+{
+    if (!_camera_definition) {
+        LogWarn() << "Error: no camera definition available yet";
+        return false;
+    }
+
+    CameraDefinition::parameter_map_t parameters;
+    _camera_definition->get_parameters(parameters, true);
+
+    auto it = parameters.find(setting_name);
+    if (it != parameters.end()) {
+        for (auto option : it->second->options) {
+            options.push_back(option->name);
+        }
+        return true;
+    } else {
+        LogErr() << "Error: setting name not found";
+        return false;
+    }
+}
+
+void CameraImpl::set_option_key_async(const std::string &setting_key,
+                                      const std::string &option_key,
+                                      const Camera::result_callback_t &callback)
+{
+    if (!_camera_definition) {
+        LogWarn() << "Error: no camera defnition available yet.";
+        if (callback) {
+            callback(Camera::Result::ERROR);
+        }
+    }
+
+    LogWarn() << "Setting an option not implemented yet";
+    if (callback) {
+        callback(Camera::Result::DENIED);
+    }
+}
+
+void CameraImpl::get_option_key_async(const std::string &setting_key,
+                                      const Camera::get_option_callback_t &callback)
+{
+    if (_camera_definition) {
+        LogWarn() << "Error: no camera defnition available yet.";
+        if (callback) {
+            callback(Camera::Result::ERROR, "");
+        }
```

Comment:
I am not sure. Don't we need to return from here ? It seems falling through.

Reply:
Very good catch, thank you!